File: py_app/visualization/bev.py
import numpy as np
import cv2
from typing import Dict, List, Optional, Tuple

from py_app.core.datatypes import Detection, Prediction, CameraConfig, BevConfig
from py_app.core.geometry import project_pixel_to_ground_undistorted, project_world_to_bev
from py_app.core.depth_utils import compute_bev_points_from_depth
from py_app.core.risk import risk_to_color


def build_bev_background(bev_config: BevConfig) -> np.ndarray:
    bev = np.zeros((bev_config.bev_size, bev_config.bev_size, 3), dtype=np.uint8)
    bev[:] = (22, 22, 22)

    cx = bev_config.bev_size // 2
    y_base = bev_config.bev_size - 20
    bonnet_w = int(0.34 * bev_config.bev_size)
    bonnet_h = int(0.08 * bev_config.bev_size)

    overlay = bev.copy()
    cv2.ellipse(overlay, (cx, y_base), (bonnet_w // 2, bonnet_h), 0, 180, 360, (255, 255, 255), 2, cv2.LINE_AA)
    cv2.ellipse(overlay, (cx, y_base), (bonnet_w // 2, bonnet_h), 0, 180, 360, (55, 55, 55), 2, cv2.LINE_AA)

    inner_w = int(bonnet_w * 0.72)
    inner_h = int(bonnet_h * 0.58)
    cv2.ellipse(overlay, (cx, y_base + 1), (inner_w // 2, inner_h), 0, 180, 360, (115, 115, 115), 1, cv2.LINE_AA)

    cv2.addWeighted(overlay, 0.92, bev, 0.08, 0, bev)
    return bev


def compute_ground_distance(
    frame_bgr: np.ndarray,
    det: Detection,
    camera_config: CameraConfig,
    bev_config: BevConfig,
) -> Optional[Tuple[float, float, float, float, float]]:
    h, w = frame_bgr.shape[:2]
    x1, y1, x2, y2 = det.bbox_xyxy_norm

    foot_x = float(np.clip((x1 + x2) * 0.5, 0.0, 1.0) * (w - 1))
    box_h = max(0.0, y2 - y1)
    foot_y_norm = min(1.0, y2 + 0.08 * box_h)
    foot_y = float(foot_y_norm * (h - 1))

    ground_pt = project_pixel_to_ground_undistorted(
        u=foot_x,
        v=foot_y,
        K=camera_config.K,
        D=camera_config.D,
        cam_height_m=camera_config.cam_height_mm / 1000.0,
        cam_pitch_deg=camera_config.cam_pitch_deg,
    )
    if ground_pt is None:
        return None

    x_m, z_m = ground_pt

    if abs(x_m) > bev_config.bev_half_width_m or z_m < 0.0 or z_m > bev_config.max_range_m:
        return None

    z_m -= 3
    dist_m = float(np.sqrt(x_m * x_m + z_m * z_m))
    return foot_x, foot_y, x_m, z_m, dist_m


def draw_depth_on_bev(
    bev: np.ndarray,
    depth: np.ndarray,
    camera_config,
    bev_config,
    depth_scale_m_per_unit: Optional[float],
) -> np.ndarray:

    pts_world, vals = compute_bev_points_from_depth(
        depth=depth,
        K=camera_config.K,
        half_width_m=bev_config.bev_half_width_m,
        max_range_m=bev_config.max_range_m,
        scale_m_per_unit=depth_scale_m_per_unit,
        sample_step=bev_config.depth_sample_step,
        max_points=bev_config.depth_max_points,
        min_row_frac=0.45,
    )

    if pts_world.shape[0] == 0:
        return bev

    x_m = pts_world[:, 0]
    z_m = pts_world[:, 1]
    z_m = np.clip(z_m, 0.0, bev_config.max_range_m)

    x_norm = (x_m + bev_config.bev_half_width_m) / (2.0 * bev_config.bev_half_width_m)
    y_norm = 1.0 - (z_m / bev_config.max_range_m)

    px = np.clip((x_norm * (bev_config.bev_size - 1)).astype(np.int32), 12, bev_config.bev_size - 12)
    py = np.clip((y_norm * (bev_config.bev_size - 1)).astype(np.int32), 0, bev_config.bev_size - 1)

    occupancy = np.zeros((bev_config.bev_size, bev_config.bev_size), dtype=np.float32)
    np.maximum.at(occupancy, (py, px), vals)

    mask = occupancy > 0
    intensity = (45 + 110 * occupancy[mask]).astype(np.uint8)

    bev[mask] = np.stack([intensity] * 3, axis=1)

    return bev


def render_bev(
    frame_bgr: np.ndarray,
    detections: List[Detection],
    predictions: Dict[int, Prediction],
    camera_config: CameraConfig,
    bev_config: BevConfig,
    depth: Optional[np.ndarray] = None,
    depth_scale_m_per_unit: Optional[float] = None,
    alpha: float = 0.1,
) -> np.ndarray:
    bev = build_bev_background(bev_config)

    if not hasattr(render_bev, "previous_positions"):
        render_bev.previous_positions = {}

    if depth is not None:
        bev = draw_depth_on_bev(
            bev=bev,
            depth=depth,
            camera_config=camera_config,
            bev_config=bev_config,
            depth_scale_m_per_unit=depth_scale_m_per_unit,
        )

        if depth_scale_m_per_unit is not None:
            cv2.putText(
                bev,
                f"depth scale={depth_scale_m_per_unit:.3f} m/unit",
                (14, bev_config.bev_size - 16),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.45,
                (180, 180, 180),
                1,
                cv2.LINE_AA,
            )

    for det in detections:
        info = compute_ground_distance(frame_bgr, det, camera_config, bev_config)
        if info is None:
            continue

        _foot_x, _foot_y, x_m, z_m, dist_m = info
        bev_x, bev_y = project_world_to_bev(
            x_m,
            z_m,
            bev_config.bev_size,
            bev_config.bev_half_width_m,
            bev_config.max_range_m,
        )

        if det.track_id in render_bev.previous_positions:
            prev_bev_x, prev_bev_y = render_bev.previous_positions[det.track_id]
            bev_x = alpha * bev_x + (1 - alpha) * prev_bev_x
            bev_y = alpha * bev_y + (1 - alpha) * prev_bev_y

        render_bev.previous_positions[det.track_id] = (bev_x, bev_y)

        bev_y -= int((3.0 / bev_config.max_range_m) * (bev_config.bev_size - 1))

        pred = predictions.get(det.track_id)

        if pred is None:
            color = (255, 255, 255)
            label = f"{dist_m:.1f}m"
        else:
            if pred.distance_m is None:
                pred.distance_m = dist_m
            color = risk_to_color(pred.risk_score)
            label = f"{pred.distance_m:.1f}m r={pred.risk_score:.2f}"

        cv2.circle(bev, (int(bev_x), int(bev_y)), 4, color, 2, cv2.LINE_AA)
        tx = min(bev_config.bev_size - 150, int(bev_x) + 12)
        ty = max(18, int(bev_y) - 8)
        cv2.putText(bev, label, (tx, ty), cv2.FONT_HERSHEY_SIMPLEX, 0.45, color, 1, cv2.LINE_AA)

    return bev

File: ros2_ws/src/pedestrian_visualizer_cpp/src/bev.cpp
#include "pedestrian_visualizer_cpp/bev.hpp"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <unordered_map>

namespace pedestrian_visualizer_cpp
{
    static cv::Scalar risk_to_color(float risk)
    {
        risk = std::clamp(risk, 0.0f, 1.0f);
        int r = static_cast<int>(255.0f * risk);
        int g = static_cast<int>(255.0f * (1.0f - risk));
        return cv::Scalar(0, g, r);
    }

    static cv::Mat build_bev_background(const BevConfig& bev_config)
    {
        cv::Mat bev(bev_config.bev_size, bev_config.bev_size, CV_8UC3, cv::Scalar(22, 22, 22));

        int cx = bev_config.bev_size / 2;
        int y_base = bev_config.bev_size - 20;
        int bonnet_w = static_cast<int>(0.34f * bev_config.bev_size);
        int bonnet_h = static_cast<int>(0.08f * bev_config.bev_size);

        cv::Mat overlay = bev.clone();

        cv::ellipse(
            overlay,
            cv::Point(cx, y_base),
            cv::Size(bonnet_w / 2, bonnet_h),
            0, 180, 360,
            cv::Scalar(255, 255, 255),
            2,
            cv::LINE_AA);

        cv::ellipse(
            overlay,
            cv::Point(cx, y_base),
            cv::Size(bonnet_w / 2, bonnet_h),
            0, 180, 360,
            cv::Scalar(55, 55, 55),
            2,
            cv::LINE_AA);

        int inner_w = static_cast<int>(bonnet_w * 0.72f);
        int inner_h = static_cast<int>(bonnet_h * 0.58f);

        cv::ellipse(
            overlay,
            cv::Point(cx, y_base + 1),
            cv::Size(inner_w / 2, inner_h),
            0, 180, 360,
            cv::Scalar(115, 115, 115),
            1,
            cv::LINE_AA);

        cv::addWeighted(overlay, 0.92, bev, 0.08, 0.0, bev);
        return bev;
    }

    static cv::Mat make_camera_matrix(const CameraConfig& cam)
    {
        return (cv::Mat_<float>(3, 3) <<
            cam.fx, 0.0f, cam.cx,
            0.0f, cam.fy, cam.cy,
            0.0f, 0.0f, 1.0f);
    }

    static cv::Mat make_distortion_vector(const CameraConfig& cam)
    {
        if (cam.dist_coeffs.empty()) {
            return cv::Mat::zeros(4, 1, CV_32F);
        }

        cv::Mat D(static_cast<int>(cam.dist_coeffs.size()), 1, CV_32F);
        for (size_t i = 0; i < cam.dist_coeffs.size(); ++i) {
            D.at<float>(static_cast<int>(i), 0) = cam.dist_coeffs[i];
        }
        return D;
    }

    // Visualization-only placement helper.
    static bool compute_ground_point(
        const cv::Mat& frame_bgr,
        const VizTrack& tr,
        const CameraConfig& cam,
        const BevConfig& bev_cfg,
        float& x_m,
        float& z_m)
    {
        const int h = frame_bgr.rows;
        const int w = frame_bgr.cols;

        float foot_x_norm = std::clamp((tr.x1 + tr.x2) * 0.5f, 0.0f, 1.0f);
        float foot_x = foot_x_norm * static_cast<float>(w - 1);

        float box_h = std::max(0.0f, tr.y2 - tr.y1);
        float foot_y_norm = std::min(1.0f, tr.y2 + 0.08f * box_h);
        float foot_y = foot_y_norm * static_cast<float>(h - 1);

        const cv::Mat K = make_camera_matrix(cam);
        const cv::Mat D = make_distortion_vector(cam);

        std::vector<cv::Point2f> src{ cv::Point2f(foot_x, foot_y) };
        std::vector<cv::Point2f> undist;
        cv::undistortPoints(src, undist, K, D);

        if (undist.empty()) {
            return false;
        }

        float x_n = undist[0].x;
        float y_n = undist[0].y;

        float pitch = cam.cam_pitch_deg * static_cast<float>(M_PI) / 180.0f;
        float beta = std::atan(y_n);

        float denom = std::tan(pitch + beta);
        if (std::abs(denom) < 1e-6f) {
            return false;
        }

        z_m = cam.cam_height_m / denom;
        if (z_m <= 0.0f) {
            return false;
        }

        x_m = x_n * z_m;

        if (std::abs(x_m) > bev_cfg.bev_half_width_m ||
            z_m < 0.0f || z_m > bev_cfg.max_range_m) {
            return false;
        }

        z_m -= 3.0f;
        return true;
    }

    static cv::Point world_to_bev(
        float x_m,
        float z_m,
        const BevConfig& cfg)
    {
        float x_norm = (x_m + cfg.bev_half_width_m) / (2.0f * cfg.bev_half_width_m);
        float y_norm = 1.0f - (z_m / cfg.max_range_m);

        int px = std::clamp(static_cast<int>(x_norm * (cfg.bev_size - 1)), 12, cfg.bev_size - 12);
        int py = std::clamp(static_cast<int>(y_norm * (cfg.bev_size - 1)), 0, cfg.bev_size - 1);
        return cv::Point(px, py);
    }

    cv::Mat render_bev(
        const cv::Mat& frame_bgr,
        const std::vector<VizTrack>& tracks,
        const std::unordered_map<int, VizPrediction>& predictions,
        const CameraConfig& camera_config,
        const BevConfig& bev_config)
    {
        cv::Mat bev = build_bev_background(bev_config);

        static std::unordered_map<int, cv::Point2f> previous_positions;
        const float alpha = 0.1f;

        for (const auto& tr : tracks) {
            float x_m = 0.0f;
            float z_m = 0.0f;

            if (!compute_ground_point(frame_bgr, tr, camera_config, bev_config, x_m, z_m)) {
                continue;
            }

            cv::Point raw_p = world_to_bev(x_m, z_m, bev_config);
            cv::Point2f p(static_cast<float>(raw_p.x), static_cast<float>(raw_p.y));

            auto prev_it = previous_positions.find(tr.track_id);
            if (prev_it != previous_positions.end()) {
                p.x = alpha * p.x + (1.0f - alpha) * prev_it->second.x;
                p.y = alpha * p.y + (1.0f - alpha) * prev_it->second.y;
            }
            previous_positions[tr.track_id] = p;

            cv::Point draw_p(static_cast<int>(p.x), static_cast<int>(p.y));
            draw_p.y -= static_cast<int>((3.0f / bev_config.max_range_m) * (bev_config.bev_size - 1));
            draw_p.y = std::clamp(draw_p.y, 0, bev_config.bev_size - 1);

            cv::Scalar color(255, 255, 255);
            char label[128];

            auto it = predictions.find(tr.track_id);
            if (it == predictions.end()) {
                std::snprintf(label, sizeof(label), "ID=%d", tr.track_id);
            }
            else {
                const auto& pred = it->second;
                color = risk_to_color(pred.risk_score);

                if (pred.distance_m.has_value()) {
                    std::snprintf(label, sizeof(label), "%.1fm r=%.2f",
                        *pred.distance_m, pred.risk_score);
                }
                else {
                    std::snprintf(label, sizeof(label), "r=%.2f", pred.risk_score);
                }
            }

            cv::circle(bev, draw_p, 4, color, 2, cv::LINE_AA);
            cv::putText(
                bev,
                label,
                cv::Point(std::min(bev.cols - 150, draw_p.x + 12), std::max(18, draw_p.y - 8)),
                cv::FONT_HERSHEY_SIMPLEX,
                0.45,
                color,
                1,
                cv::LINE_AA);
        }

        return bev;
    }
}